pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,257 @@
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	var pgInfo1 PodGroupInfo
+	var pgInfo2 PodGroupInfo
+	var ok bool
+	podGroupName1, min1 := cs.checkPod(pod1)
+	podGroupName2, min2 := cs.checkPod(pod2)
+	if podGroupName1 != "" {
+		pgInfo1, ok = cs.PodGroupInfos[podGroupName1]
+		if !ok {
+			pgInfo1 = PodGroupInfo{
+				InitialTimestamp: pod1.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min1,
+			}
+			cs.PodGroupInfos[podGroupName1] = pgInfo1
+		}
+	}
+	if podGroupName2 != "" {
+		pgInfo2, ok = cs.PodGroupInfos[podGroupName2]
+		if !ok {
+			pgInfo2 = PodGroupInfo{
+				InitialTimestamp: pod2.CreationTimestamp.Time,
+				Id:               cs.GetId(),
+				MinAvailable:     min2,
+			}
+			cs.PodGroupInfos[podGroupName2] = pgInfo2
+		}
+	}
+
+	if podGroupName1 != "" && podGroupName2 != "" {
+		if pgInfo1.InitialTimestamp.Equal(pgInfo2.InitialTimestamp) {
+			return pgInfo1.Id < pgInfo2.Id
+		} else {
+			return pgInfo1.InitialTimestamp.Before(pgInfo2.InitialTimestamp)
+		}
+	} else if podGroupName1 != "" {
+		return pgInfo1.InitialTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+	} else if podGroupName2 != "" {
+		return podInfo1.InitialAttemptTimestamp.Before(pgInfo2.InitialTimestamp)
+	}
+
+	return podInfo1.InitialAttemptTimestamp.Before(podInfo2.InitialAttemptTimestamp)
+}
+
+func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, p *v1.Pod) *framework.Status {
+	podGroupName, minAvailable := cs.checkPod(p)
+	if podGroupName == "" || minAvailable == 0 {
+		return framework.NewStatus(framework.Success, "")
+	}
+
+	total := cs.calculateTotalPods(podGroupName)
+	if total < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in PreFilter: %d",
+			podGroupName, minAvailable, total)
+		return framework.NewStatus(framework.Unschedulable, "less than minAvailable")
+	}
+
+	return framework.NewStatus(framework.Success, "")
+}
+
+func (cs *Coscheduling) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Permit is the functions invoked by the framework at "permit" extension point.
+func (cs *Coscheduling) Permit(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) (*framework.Status, time.Duration) {
+	podGroupName, minAvailable := cs.checkPod(p)
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	running := cs.calculateRunningPods(podGroupName)
+	waiting := cs.calculateWaitingPods(podGroupName)
+
+	if running+waiting+1 < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			podGroupName, minAvailable, running, waiting)
+		return framework.NewStatus(framework.Wait, ""), 10 * time.Second
```

Comment:
I think of a generation algorithm about the dynamic timeout. We know that a group of pods is exactly the same. When the cluster node is unchanged, we can assume that the scheduling time of each pod is the same, then we can get a formula
(minAvailable - waitingPods - 1) * schedulingTime, where schedulingTime is the time that pod joins the scheduling queue to run to the permit plug-in. For example, if the minAvailable of a job is 5 and the time to schedule a pod is 10ms, then the waiting time for the first pod is (5-0-1) * 10ms, the waiting time for the second pod is (5-1-1) * 10ms, and so on for the rest of the pods. Considering network, load, etc., it is possible to add a random time to the timeout. WDYT?

Reply:
thanks, I prefer the first version to use `minAvailable-running-waiting-1/minAvailable` * SomeTime (A constant value) as the timeout. I think using the schedulingTime(use the time.Now() - PodInfo.Timestamp) is a good option, Maybe we need more estimate to check it.

Reply:
SGTM